projects/rocr-runtime/runtime/hsa-runtime/core/inc/counted_queue_manager.h

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright © Advanced Micro Devices, Inc., or its affiliates. 
+ * 
+ * SPDX-License-Identifier: MIT
+ */
+
+#ifndef HSA_RUNTME_CORE_INC_COUNTED_QUEUE_MANAGER_H_
+#define HSA_RUNTME_CORE_INC_COUNTED_QUEUE_MANAGER_H_
+
+
+#include "hsa.h"
+#include "hsa_ext_amd.h"
+#include "core/inc/agent.h"
+#include "core/inc/queue.h"
+#include "core/inc/runtime.h"
+
+#include "map"
+
+namespace rocr {
+namespace core {
+
+// Wrapper for real hardware queue tracking and its references
+// Each entry in queue pool is of this type
+struct HardwareQueue {
```

Comment:
I don't think we need a new struct to represent a HW queue. You can use the core::Queue class in (queue.h). 

Then CountedQueue would have a pointer to a core::Queue* 

We can extend the use core::Queue class to have to have a members for use_count and priority.